adobe/spectrum-web-components
# Fix overlay open close symmetry

## Description

This fixes some overlapping race conditions in the way we track which overlays are currently open. Opening and closing an overlay is an async operation, but we use the existence of `closeClickOverlay`, `closeHoverOverlay`, etc to synchronously manage overlay state.

This PR converts the close functions to `Promises` and centralizes their async usage into `closeAllOverlays`, eliminating several `await`s and `delete`s.

## Related issue(s)

Resolves https://github.com/adobe/spectrum-web-components/issues/1834

## Motivation and context

An overlay trigger that has both hover and click content should be able to open and close by repeatedly being clicked.

## How has this been tested?

1. Go to the first example in https://opensource.adobe.com/spectrum-web-components/components/overlay-trigger
2. Hover with your mouse and wait for the tooltip
3. Click repeatedly
4. Notice that the overlay does not close
---
1. Go to the first example in https://fix-overlay-open-close-symmetry--spectrum-web-components.netlify.app/components/overlay-trigger
2. Hover with your mouse and wait for the tooltip
3. Click repeatedly
4. Notice that the overlay closes
---
1. `yarn test`

## Types of changes


## Checklist


## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.
